python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,157 @@
+"""checker for variables that might not be defined in the program.
+"""
+from typing import Union
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+from python_ta.cfg.graph import CFGBlock
+from astroid.node_classes import NodeNG
+from typing import Set, List
+
+
+class PossiblyUndefinedChecker(BaseChecker):
+
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'possibly_undefined'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9969': ('This variable might not be defined when this statement is executed.',
+                      'possibly-undefined',
+                      'Reported when a statement uses a variable that might not be assigned.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    possibly_undefined = set()
+
+    @check_messages('possibly-undefined')
+    def visit_name(self, node):
+        if self._check_possibly_undefined(node):
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _check_possibly_undefined(self, node: astroid.Name) -> bool:
+        """Returns True if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        return node in self.possibly_undefined
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `possibly_undefined` if it might not be defined.
+        """
+        facts = {}
+        blocks = self._get_blocks_po(node)
+
+        all_assigns = self._get_assigns(blocks)
+        for block in blocks:
+            facts[block] = {}
+            facts[block]['out'] = all_assigns.copy()
+
+        worklist = blocks[:]
```

Comment:
As far as I can tell `blocks` isn't used after this, which means you shouldn't need to create a copy here.